### DISCUSSION TITLE
After inserting the data, do I need to build the index again?

### DISCUSSION THREAD

**Discussion by ingale726** [milvus-io/milvus#22280] [Q&A and General discussion]
After creating the collection, I first create the index and then insert the data. After inserting the data, do I need to create the index again? When I inserted 140 million data and then built the index, it was very time consuming, I wondered if it was necessary?

**Comment by yhmo** [comment]
Two ways to create index:

- create an empty collection, then specify an index type, then insert data continually, the index node will automatically build index for sealed segments.
steps:
1. create_collection()
2. create_index()
3. insert(), insert(), insert() ......

- create an empty collection, after all data is inserted, call flush() to seal all segments, then call create_index() to build index for all segments
steps:
1. create_collection()
2. insert(), insert(), insert() ......
3. create_index()


Index operation only happens on the segment which is "sealed". If your data is already inserted, you want to make sure all the segments are indexed, you can call flush() and create_index() to do the work. The create_index() will be blocked until all the segments are indexed.
For streaming data scenario, usually follow the way_1, just let milvus to automatically build index internally.

**Reply by ingale726** [reply]
I used method 1, which is to build an empty collection, index it, insert data, and then load it, but I found that the load was very slow, about the same time it took to build the index. I think it creates an index again, right?

**Reply by yhmo** [reply]
Loading and index building are in different threads/processes, they don't block each other.
"load was very slow", the possible reasons:
- storage system bandwidth is poor
- huge dataset, lot of data files need to be loaded
